src/batch/src/monitor/stats.rs

```diff
@@ -0,0 +1,398 @@
+            ($({ $metric:ident, $type:ty },)*) => {
+                $(
+                    mfs.extend(self.$metric.collect());
+                )*
+            };
+        }
+        for_all_task_metrics!(collect);
+
+        // TODO: Every time we execute it involving get the lock, here maybe a bottleneck.
+        self.clean_metrics();
+
+        mfs
+    }
+}
+
+macro_rules! for_all_executor_metrics {
+    ($macro:ident) => {
+        $macro! {
+            { exchange_recv_row_number,  GenericCounterVec<AtomicU64>, GenericCounter<AtomicU64>},
+            { row_seq_scan_next_duration, HistogramVec , Histogram},
+        }
+    };
+}
+
+macro_rules! def_executor_metrics {
+    ($( { $metric:ident, $type:ty, $_t:ty }, )*) => {
+        #[derive(Clone)]
+        pub struct BatchExecutorMetrics {
+            descs: Vec<Desc>,
+            delete_task: Arc<Mutex<Vec<TaskId>>>,
+            register_labels: Arc<Mutex<HashMap<TaskId, Vec<Box<dyn Fn() + Send>>>>>,
+            $( pub $metric: $type, )*
+        }
+    };
+}
+
+for_all_executor_metrics!(def_executor_metrics);
+
+impl BatchExecutorMetrics {
+    pub fn new(register: Registry) -> Self {
+        let executor_labels = vec!["query_id", "stage_id", "task_id", "executor_id"];
+        let mut descs = Vec::with_capacity(2);
+
+        let mut custom_labels = executor_labels.clone();
+        custom_labels.extend_from_slice(&["source_query_id", "source_stage_id", "source_task_id"]);
+        let exchange_recv_row_number = IntCounterVec::new(
+            opts!(
+                "batch_exchange_recv_row_number",
+                "Total number of row that have been received from upstream source",
+            ),
+            &custom_labels,
+        )
+        .unwrap();
+        descs.extend(exchange_recv_row_number.desc().into_iter().cloned());
+
+        let task_row_seq_scan_next_duration = HistogramVec::new(
+            HistogramOpts::new(
+                "batch_row_seq_scan_next_duration",
+                "Time spent deserializing into a row in cell based table.",
+            )
+            .buckets(exponential_buckets(0.0001, 2.0, 20).unwrap()),
+            &executor_labels,
+        )
+        .unwrap();
+        descs.extend(task_row_seq_scan_next_duration.desc().into_iter().cloned());
+
+        let metrics = Self {
+            descs,
+            delete_task: Arc::new(Mutex::new(Vec::new())),
+            exchange_recv_row_number,
+            row_seq_scan_next_duration: task_row_seq_scan_next_duration,
+            register_labels: Arc::new(Mutex::new(HashMap::new())),
+        };
+        register.register(Box::new(metrics.clone())).unwrap();
+        metrics
+    }
+
+    /// Create a new `BatchTaskMetrics` instance used in tests or other places.
+    pub fn for_test() -> Self {
+        Self::new(prometheus::Registry::new())
+    }
+
+    fn clean_metrics(&self) {
+        let delete_labels = {
+            let mut delete_task = self.delete_task.lock();
```

Comment:
Yes. And another optimization is to call clean_metrics with a interval to avoid acquire lock every collect. But not sure we should do it immediately. Maybe wait for a benchmark test to decide whether it have a big impact. 🤔